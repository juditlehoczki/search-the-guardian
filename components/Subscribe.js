import React, { Component } from "react";
import { View, Text, TextInput, TouchableOpacity, Alert } from "react-native";
import s from "../styles.js";

class Subscribe extends Component {
  state = { emailInput: "", emailError: null };

  handleChange(value) {
    this.setState({
      emailInput: value,
    });
    this.validate(value);
  }

  handleSubmit() {
    const { emailInput, emailError } = this.state;
    if (emailInput === "") {
      this.setState({ emailError: "Enter your email address first." });
    } else if (emailError === null) {
      this.setState({ emailInput: "" });
      Alert.alert(
        "You have successfully subscribed.",
        "Thank you!",
        [
          {
            text: "Close",
            style: "cancel",
          },
        ],
        { cancelable: false }
      );
    }
  }

  validate(value) {
    const regex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    regex.test(value)
      ? this.setState({ emailError: null })
      : this.setState({ emailError: "Please enter a valid email address." });
  }

  render() {
    const { emailInput, emailError } = this.state;
    return (
      <View>
        <Text style={s.sectionTitle}>
          Want more? Subscribe to our weekly newsletter:
        </Text>
        {emailError && <Text style={s.errorMsg}>{emailError}</Text>}
        <View style={s.inputContainer}>
          <Text style={s.icon}>✉️</Text>
          <TextInput
            style={[emailError ? s.inputFieldInvalid : s.inputField]}
            value={emailInput}
            placeholder="Email address..."
            keyboardType="email-address"
            onChangeText={(value) => {
              this.handleChange(value.toLowerCase());
            }}
          />
          <TouchableOpacity
            style={s.button}
            onPress={() => {
              this.handleSubmit();
            }}
          >
            <Text style={s.buttonText}>Go!</Text>
          </TouchableOpacity>
        </View>
      </View>
    );
  }
}

export default Subscribe;
