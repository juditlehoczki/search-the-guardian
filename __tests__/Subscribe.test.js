import React from "react";
import "react-native";
import renderer from "react-test-renderer";

import Subscribe from "../components/Subscribe.js";

describe("Subscribe Component", () => {
  beforeEach(() => {
    fakeHandleChange = jest.fn();
    fakeHandleSubmit = jest.fn();
    fakeValidate = jest.fn();
  });
  test("Snapshot - it renders correctly", () => {
    const snapshot = renderer.create(<Subscribe />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  test("handleChange method works as expected - sets the emailInput state correctly", () => {
    const fakeSubscribe = renderer.create(<Subscribe />).getInstance();
    fakeSubscribe.handleChange("lehoczki.judit@gmail.com");
    expect(fakeSubscribe.state.emailInput).toEqual("lehoczki.judit@gmail.com");
  });
});
