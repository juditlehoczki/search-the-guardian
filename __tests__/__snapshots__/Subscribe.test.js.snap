// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscribe Component Snapshot - it renders correctly 1`] = `
<View>
  <Text
    style={
      Object {
        "backgroundColor": "#54C1B7",
        "color": "white",
        "fontWeight": "bold",
        "marginVertical": 15,
        "paddingLeft": 15,
        "paddingVertical": 5,
      }
    }
  >
    Want more? Subscribe to our weekly newsletter:
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "fontSize": 20,
          "marginLeft": 5,
        }
      }
    >
      ✉️
    </Text>
    <TextInput
      allowFontScaling={true}
      keyboardType="email-address"
      onChangeText={[Function]}
      placeholder="Email address..."
      rejectResponderTermination={true}
      style={
        Object {
          "borderColor": "#D1529E",
          "borderRadius": 5,
          "borderStyle": "dotted",
          "borderWidth": 1,
          "color": "#0F406E",
          "flex": 8,
          "marginLeft": 5,
          "marginVertical": 5,
          "padding": 15,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#D1529E",
          "borderRadius": 15,
          "flex": 1,
          "justifyContent": "center",
          "margin": 5,
          "opacity": 1,
          "padding": 5,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontWeight": "bold",
          }
        }
      >
        Go!
      </Text>
    </View>
  </View>
</View>
`;
